linode/manager
# upcoming: [DI-19610] - Added CloudPulse widget to show graph for different metrices

## Description 📝
Added Cloud Pulse widgets to visualize the graphs for different metrices for the selected service type

## Changes  🔄
List any change relevant to the reviewer.
1. Dashboard Component is added to fetch the metric definitions data & load widgets for all the metric definitions available.
2. CloudPulse Widget component is added to load the graphs & various filters available for each metrics.
3. Aggregate Function component is added to show widget data based on different aggregate functions.
4. Interval component is added to show widget data at selected time interval.
5. Zoomer component is added to zoom-in or zoom-out individual widgets.
6. Added logic to update widget data to preferences 
7. A refresh button is added in global filters to manually refresh the metrics data. (Functionality will be added in future PRs).

## Target release date 🗓️
Target Release date: 22nd July 2024

## Preview 📷
**Include a screenshot or screen recording of the change**

:bulb: Use `<video src="" />` tag when including recordings in table.

| Before  | After   |
| ------- | ------- |
| ![Screenshot 2024-07-15 at 5 27 28 PM](https://github.com/user-attachments/assets/5c775018-af9d-4089-b8fb-1a76d24749dc) | ![Screenshot 2024-07-15 at 5 35 33 PM](https://github.com/user-attachments/assets/26706ea0-4739-4cb0-bf3c-348df77c147f) |
|![Screenshot 2024-07-19 at 12 44 42 PM](https://github.com/user-attachments/assets/36efb566-fa4b-4f8d-b448-8155446c7ed4)|![Screenshot 2024-07-19 at 12 43 41 PM](https://github.com/user-attachments/assets/f86a0cf4-b582-41c3-b43f-9e7602dc6d79)|
|![Screenshot 2024-07-19 at 12 43 58 PM](https://github.com/user-attachments/assets/3933af4a-682f-4f37-88f4-f378bf942f4f)|![Screenshot 2024-07-19 at 12 44 21 PM](https://github.com/user-attachments/assets/460f258a-c8fc-4c86-b817-999077506251)|


## How to test 🧪

1. Enable mock services as some endpoints are not ready to use.
2. Since the PR has only structural view of the widgets & not its actual functionality to load the data. You can play around with different drop down & zoom-in /zoom-out icon to check the working.
3. Preferences are also added to update widget changes, so you can verify from network tab that on any drop down change or zoom-icon click, it'll do a network call to preferences api.

Note: Currently a dummy graph is added to showcase how complete widget will look like & actual graph implementation will be added in future PRs


## As an Author I have considered 🤔

*Check all that apply*
